### ISSUE TITLE
running mtf in a container results in odd traceback

### ISSUE THREAD

**Issue by TomasTomecek** [fedora-modularity/meta-test-family#16]
```
Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado/core/test.py:576
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/moduleframework/module_framework.py", line 1121, in setUp
    return self.backend.setUp()
  File "/usr/lib/python2.7/site-packages/moduleframework/module_framework.py", line 286, in setUp
    self.__prepareContainer()
  File "/usr/lib/python2.7/site-packages/moduleframework/module_framework.py", line 321, in __prepareContainer
    service_manager = service.ServiceManager()
  File "/usr/lib/python2.7/site-packages/avocado/utils/service.py", line 774, in service_manager
    internal_service_manager = _service_managers[get_name_of_init(run)]
KeyError: 'make'
```

code from [here](https://github.com/container-images/container-image-template/pull/32)

**Comment by jscotka** [comment]
Hi,
you are trying to run tests inside container?
If yes, as I mentioned, MTF needs some improvements to avoid any calls what are not permitted:
for example it tries to start docker via systemd if not started (via avocado utils.service tool) so that it needs to contain something what found if docker is already running, or some option to disable all setup parts of MTF like: 
It tries to install docker in case not installed: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/module_framework.py#L298
and it also tries to add alternative repo and add insecure registry item + restart docker in function: 
https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/module_framework.py#L307


I suggest to add some variable to disable these setup parts and expect that system is fully ready for testing with proper sources

**Comment by TomasTomecek** [comment]
> I suggest to add some variable to disable these setup parts and expect that system is fully ready for testing with proper sources

You don't need variable for that. If `/var/run/docker.sock` is present, you're good to go.

**Comment by phracek** [comment]
This is a bit tricky. You are trying to run mtf in docker container, which is fine, but MODULE=docker will try to run docker inside docker. 
MTF loads ServiceManager https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/module_framework.py#L321

The service Manager is load here https://github.com/avocado-framework/avocado/blob/master/avocado/utils/service.py#L789 and tries to get name of init system like SysInitV or systemd https://github.com/avocado-framework/avocado/blob/master/avocado/utils/service.py#L774
Defined here https://github.com/avocado-framework/avocado/blob/master/avocado/utils/service.py#L774

Unfortunately in Dockerfile is not present SysInitV or systemd.

function which detects those two kind of init systems is https://github.com/avocado-framework/avocado/blob/master/avocado/utils/service.py#L426 respectively here https://github.com/avocado-framework/avocado/blob/master/avocado/utils/service.py#L398
it checks /proc/1/exe which is in your case make.

**Comment by TomasTomecek** [comment]
I genuinely don't understand why there is a code a like this in mtf: https://github.com/fedora-modularity/meta-test-family/blob/master/moduleframework/module_framework.py#L322

I personally don't like when a test runner/framework is messing with my packages and system services - it should try to run selected tests in an environment I'm providing, not messing with my system. Do you know why? Because when there is a use case to run tests in a container, suddenly it's a big problem.

**Comment by phracek** [comment]
You are right, this will cause troubles in future. This has to be repaired as soon as possible.
It would be nice to detect whether MTF is running inside container and do not run docker at all.

**Comment by phracek** [comment]
You are right, this will cause troubles in future. This has to be repaired as soon as possible.
It seems you hit a big point in MTF. User should be able to test their use cases in container.
Therefore I suggest, detect if MTF is running inside container and if so, run tests without docker engine.

Otherwise I have no idea how to run tests with MTF in docker container as soon as you specify MODULE=docker.

@TomasTomecek What is your suggestion? And thanks for really good point.

**Comment by TomasTomecek** [comment]
> @TomasTomecek What is your suggestion?

I already tried to provide my suggestion:

> it should try to run selected tests in an environment I'm providing, not messing with my system.

The way I see it is that the code `service_manager.start('docker')` is just a fix to an issue when docker is not running. What happens when the docker service can't be started? Will you try to write a code to fix such issue? What if network is down? Will you try to restart NetworkManager? I could continue here *ad absurdum*. My opinion here is that some issues can be fixed or being worked around and some need manual intervention from the user/operator. Hence I think that testing frameworks and test runners  should not change state of system services or change installed system packages -- you can describe those as dependencies in spec file.

I understand there are some architectural decisions made for the way mtf is suppose to work and one issue cannot change those. Hence I think we should sit down and talk about reasons of those decisions and see if we can 1) change them 2) improve them 3) or leave them be and try to figure out this issue somehow.

**Comment by ignatenkobrain** [comment]
@TomasTomecek ++

**Comment by bkabrda** [comment]
I totally agree with @TomasTomecek that the test framework shouldn't install any packages or temper with system services. As Tomas notes, the test framework just has to consider these things taken care of, but not touch them. If the framework can't detect some of these things, then it should fail with a meaningful error message that tells the user what's wrong and what may be a good way to fix it.

**Comment by phracek** [comment]
I understand, agree with both of you @TomasTomecek @bkabrda , but who will prepare an environment for testing RPM based modules, docker images, and the others? I have added and new option MODULE=only_tests which does exactly the think which you are talking about. I guess, the first reason in MTF was just write the tests and don't care about environment. Keep it simple. Two approaches.

We should wait for Jan and discuss it. I like the Jan's approach, like only write the tests and don't care about environment. Enter in config.yaml image and MTF will do all stuff behind you.

And thanks for your feedback. Really positive.

**Comment by TomasTomecek** [comment]
> but who will prepare an environment for testing RPM based modules, docker images, and the others?

On my laptop, I will prepare the environment -- start required services, install missing packages (which should be pulled as rpm dependencies anyway, right?)

In CI, we should utilize work of Stef's team: https://fedoraproject.org/wiki/Changes/InvokingTests

**Comment by phracek** [comment]
This issue will be solved by this RFE #31 . Linking issue.

**Comment by rpitonak** [comment]
Can we close this one? It's solved by #44
